feat: add new tree-view component

alenaksu opened the pull request:
This PR adds a Tree and TreeItem components that allow user to display a hierarchical list of items, expanding and collapsing the nodes that have nested items.

The `sl-tree` component implements the API to interact with the whole tree, like the selection and the keyboard navigation. It has just a default slot that can contain `sl-tree-item` components, that represent the nodes of the tree.

Each `sl-tree-item` can in turn contain other `sl-tree-item`, allowing to the user to define nested trees. A tree item with nested nodes, can be expanded and collapsed.

Closes #539

## Code example
```html
<sl-tree>
    <sl-tree-item>Node</sl-tree-item>
    <sl-tree-item>
      Parent Node
      <sl-tree-item>Child Node</sl-tree-item>
    </sl-tree-item>
</sl-tree>
```

## Appereance
![Screenshot from 2022-07-23 10-10-58](https://user-images.githubusercontent.com/12519551/180597587-c493801e-cdfb-444b-9c10-48d967940c94.png)

alenaksu committed:
feat: add new tree-view component

feat: add indeterminate state

fix: rename item role

fix: aria attributes

fix: restore hover effect

fix: indeterminate state

chore: fix lint issue

fix: address (partially) review comment

fix: minor fix

feat: add keyboard navigation

fix: dependency name

fix: keyboard selection does not update all items

chore: minor changes

chore: remove leftover

chore: update documentation

feat: add lazy loading + several fixes

fix: add aria-busy attribute

fix: improve keyboard navigation and lazy loading

chore: fix linting issue

chore: minor fixes

fix: update component styling and slot

chore: remove exportparts attribute

fix: remove button margin for safari

fix: set correct part name

feat: implement selection modes and fix accessibility issues

chore: fix linting issues

chore: update docs

fix: minor fix

fix: treeitem height style

chore: update documentation

chore: refactor implementation

chore: implement unit tests

chore: update Enter key behavior

chore: update doc

chore: update doc

claviska commented:
Thanks for submitting this. I know it's a draft, but here are some initial thoughts:

- Great job with following existing patterns and using APIs such as the animations registry! (I know these aren't well documented, but it's clear you're familiar with the codebase.)
- I like that you can nest tree items inside other tree items without having to create a new tree :)
- I like the idea of allowing checkable items for easy selection.
  - Can we move the checkbox on the inside of the tree by default? (The order could be customized by end users with `order` if we use flexbox here.) <br><img width="193" alt="CleanShot 2022-07-11 at 08 44 03@2x" src="https://user-images.githubusercontent.com/55639/178266979-4398eb98-0ca3-485b-9263-5330fe4d2c29.png">
  - Should `checkable` be called `selectable` instead?
  - Should it be on the tree or the tree item? Are all tree items always selectable, even subtrees?
  - Since we're using checkboxes for checked states, can we make clicks on the items toggle expand/collapse like [this example](https://element-plus.org/en-US/component/tree.html#default-expanded-and-default-checked)
- Tree item should probably have a loading state to facilitate lazy loading of subtrees. An example of this in the docs would also be helpful since users ask for this a lot (years ago I managed a jQuery-based tree plugin and this was a top request). It's probably preferable to use events here.
- Events/slots/dependencies/etc. need to be documented (I like to do this as I go so I don't miss anything)
- Can we remove the skeletons from the first example? It makes the example feel broken like it never loads.
- I want to make sure we get accessibility right here, so let's make sure to [follow the guidelines here](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/treeitem_role) for treeitems, e.g. "No more than one node can be selected at a time unless the tree node has `aria-multiselectable="true"` set." The Shoelace version should also announce the same way in screen readers.
- The `icon` slot should probably be `prefix` for consistency with other components.
  - If we have a `prefix` slot, a `suffix` slot will also be desirable.
  - Now that I typed that out, do we even needs slots for this or should it be up to the user to put all this into the default slot? 🤔
- We should think about what parts (and exported parts) need to be exposed

I'm sure I'll have more feedback as this progresses, but you're off to a great start here!

claviska commented:
A couple comments on styles:

Can we remove the border/padding/background on the base? It appears to be in a well which is nice for some use cases, but I think the _default_ appearance should be less opinionated, i.e. the user should be able to style the host element to add a border/background/padding should they want it.

I'm also not sure about the light gray highlight + primary text color on selected tree items. 

<img width="854" alt="CleanShot 2022-07-13 at 08 54 08@2x" src="https://user-images.githubusercontent.com/55639/178738307-54ad9fbf-f714-4e70-b00f-a2e531c45477.png">

Menu uses a primary background and white text, for example. I don't think anything in the library uses a light gray background for selections, so we might need to play with this a bit. Not sure if making it look like menu items would be too bold — maybe it's just the primary text throwing it off.

> I think the behaviour follows what is described by the ARIA spec, especially if we think about the multiple selection. I'm not sure if there is a use case where just a subset of child items need to be selectable, so I kept the implementation into the tree component for now.

That's fair.

Regarding lazy loading of nodes, if tree items had a `loading` attribute that might be sufficient to facilitate async loading as long as screen readers announce it (maybe through `aria-busy` or `aria-live`).

I see some new commits since your last message, so let me know when you're ready for the next review/steps.

alenaksu commented:
> I'm also not sure about the light gray highlight + primary text color on selected tree items.

The reason why I didn't follow the same pattern used for the menu is because I thought it would have been a bit too "flashy" when using multiple selection
![image](https://user-images.githubusercontent.com/12519551/178744891-05aee801-dbfd-41b6-904d-fb962bf42549.png)
![image](https://user-images.githubusercontent.com/12519551/178745761-1e74cf1b-476c-4b7f-bcff-feefd9ce9108.png)

alenaksu commented:
I've updated the PR:

- Removed all the styles from the base element in the tree component
- Removed the gray background when highlighting a tree item, now the hover/selected item has only the primary color set (let me know what you think about my previous comment)
- Implemented a possible solution for the lazy loading: the tree-item component has a new `lazy-loading` property that indicates that the content is not loaded yet, so it displays the expand button even if there are not children. When the user tries to expand the node, the `loading` state is set to true and the `sl-load-data` event is emitted. When the `children` slot changes (new items added), we can assume that the content has been loaded so the `loaded` state is set to false.
- Added some more documentation

I think the lazy loading is quite simple to use in this way, but I'm curious to know what you think.

claviska commented:
## Accessibility

It seems like screen reader compat is very, very close! Here are some examples I referenced using VoiceOver:

- https://dequeuniversity.com/library/aria/tree-view
- http://accessibleculture.org/articles/2013/02/not-so-simple-aria-tree-views-and-screen-readers/

One thing I noticed is that it keeps announcing "button, 1 of 1" and I think it's because of this `role="button"` highlighted below:

<img width="1069" alt="CleanShot 2022-07-14 at 09 21 03@2x" src="https://user-images.githubusercontent.com/55639/178992198-5a12201e-9e50-4f62-94a1-548b1caa6a8d.png">

It's also not currently possible to tab into the tree. It looks like a roving tab index is called for, so tabbing once lands focus in the tree and another leaves it, and subsequent tabs back into the tree should focus on the last selected tree item. From that point, arrow keys are used to navigate and expand/collapse. (FYI — neither example allows <kbd>Enter</kbd> to expand/collapse a node, and it seems like <kbd>Left|Right</kbd> are the correct keys to use here so we should probably keep those and remove <kbd>Enter</kbd>.)

## Styles

> Removed the gray background when highlighting a tree item, now the hover/selected item has only the primary color set (let me know what you think about my previous comment)

Do we even need a selected style? Mouse users don't need to see it, and keyboard users will see `:focus-visible` styles so we can use that for them and do a normal focus ring outline (see other components for reference). The only time a selection needs to be visible (aside from using the keyboard) is when there's a checkbox, in which case the checkbox serves that purpose.

Note that users should be able to add a selected state through custom styles if they want it. The part could look like `part="item item--selected"` so they'd target the latter.

## Lazy Loading

Nice approach! Can we rename `lazy-loading` to `lazy` and rename `sl-load-data` to `sl-lazy-load` so they're more tightly coupled? And maybe update the docs for the event to say something like:

> Emitted when a lazy item is selected. Use this event to asynchronously load data and append items to the tree before expanding.

Question: the docs only show the lazy demo loading once. I assumed it would load every time the node expands unless the user removes the `lazy` attribute themself. Can we make this happen? It's likely we'll have two use cases for async nodes:

- load on every expand
- load once on first expand

And it feels like too strong an opinion to have the component mutate the lazy attribute for them.

alenaksu commented:
I've got just a couple of questions:

>(FYI — neither example allows Enter to expand/collapse a node, and it seems like Left|Right are the correct keys to use here so we should probably keep those and remove Enter.)

The [ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/#keyboard-interaction-24) states that particular behaviour for the `Enter` key. Do you think it's fine to remove it?

> Do we even need a selected style?

To be honest I think it would be useful, image if you are implementing a sort of file browser, and you want to keep the selected item highlighted even if you move the focus away. 
This for example, is the file tree present in the `Changed Files` tab on the Github PRs.
![image](https://user-images.githubusercontent.com/12519551/179023665-f65afd86-4304-4374-b275-e2447bf03104.png)
Does it make sense?

claviska commented:
> The [ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/#keyboard-interaction-24) states that particular behaviour for the Enter key. Do you think it's fine to remove it?

I stand corrected. I was basing that off of the two examples, neither of which expand on Enter. Carry on! 😂 

> To be honest I think it would be useful, image if you are implementing a sort of file browser, and you want to keep the selected item highlighted even if you move the focus away.

I agree, that’s a great use case. But should the visible selection be on by default for all users? It feels a lot better to me if the tree starts without a selection, and for users who want to show one, they can easily target `::part(item--selected)` to opt in and customize it to their taste.

alenaksu commented:
PR updated:

- Solved the roving tabindex issue
- Removed the `role=button` attribute from within the component
- Renamed `lazy-loading` in `lazy` and `sl-load-data` in `sl-lazy-load`
- Changed the behaviour of the lazy loading, now will emit the load event every time until the user doesn't remove the attribute manually
- A selected item is not unstyled by default, it can be styled using the new parts `item--selected` `item--disabled`
- Added a couple of demos in the doc to demonstrate how to highlight the selected item and how to display indentation lines.

jaredcwhite commented:
There's some odd spacing/indentation going on in Safari mobile. Haven't reviewed in other browsers yet.

![FB8145E9-5C93-4C1A-8566-D4B53F904AF8](https://user-images.githubusercontent.com/658496/179312974-43c1a307-217a-468b-b844-941b77581103.jpeg)

claviska commented on src/components/tree/tree.styles.ts:
```diff
@@ -0,0 +1,11 @@
+
+  :host {
+    display: block;
+    font-size: 0;
```
I'm curious what this is for.

alenaksu replied:
I was waiting for this comment 😄 
The indentation width is using the `em` unit, so the initial font-size is set to zero and at each level of the tree-item, it is increased by `1em + var(--indentation-size)` (see [here](https://github.com/shoelace-style/shoelace/pull/823/files#diff-0bcb636b92461b0ebb1b92651ad41098694b44f40b52933b62805c406c5dbc80R87) and [here](https://github.com/shoelace-style/shoelace/pull/823/files#diff-0bcb636b92461b0ebb1b92651ad41098694b44f40b52933b62805c406c5dbc80R43))

This trick was necessary to being able to focus the full row instead of just the label.

claviska replied:
Got it. Makes sense! Probably a good place for a comment :)

claviska commented:
<img width="316" alt="CleanShot 2022-07-18 at 08 08 22@2x" src="https://user-images.githubusercontent.com/55639/179508020-9791ce64-c996-40a1-889a-907ad5cf8728.png">

As I was looking at this again, I noticed GitHub has a tree in the sidebar. It looks like clicking any expandable node (not just the chevron) will cause it to expand, while clicking non-expandable items selects them.

Is this a use case we can cover with the current API?

claviska commented:
I also like how they're focus-visible styles work. It's pretty clear what's being selected:

<img width="315" alt="CleanShot 2022-07-18 at 08 10 12@2x" src="https://user-images.githubusercontent.com/55639/179508330-975d9b44-18ca-4ee4-9022-70f9463e46e5.png">

claviska commented:
> Just thinking loud, we could make the expand/collapse action the default behaviour when clicking parent nodes, then use the selectable property to enable/disable the selection of particular nodes. But since selectable is used to display the checkbox at the moment, we'd need to introduce a new internal state for that.

Should this be done at the `<sl-tree>` level instead? That would mean less markup for users and consistent behavior throughout the entire tree. Not sure what the attribute would be called, though.

An alternative might be a cancelable event that users can opt into by listening for the `sl-selected-change` and calling `event.preventDefault()`. 🤔

claviska commented on src/components/tree-item/tree-item.ts:
```diff
@@ -0,0 +1,316 @@
+ * @since 2.0
+ * @status experimental
+ *
+ * @dependency sl-checkbox sl-spinner sl-icon-button
```
This isn't obvious at all, but these need to be on separate lines otherwise all except the first will be ignored.

claviska replied:
Also not obvious — we need to auto-import checkbox, spinner, and icon button here. See color picker's source for reference. https://github.com/shoelace-style/shoelace/blob/00f98cc5050e1c06374330ed28228fbf69aa92ae/src/components/color-picker/color-picker.ts#L8-L13

alenaksu replied:
fixed [6413915](https://github.com/shoelace-style/shoelace/pull/823/commits/641391589df9c9fbd8c598caf6e8e5afe21a1a71)

claviska commented on src/components/tree/tree.ts:
```diff
@@ -0,0 +1,248 @@
+ * @since 2.0
+ * @status experimental
+ *
+ * @dependency sl-tree-item
```
I believe we can remove this. We only need to label dependencies that are used within a component's template. Since `<sl-tree-item>` is slotted in by the user, they're expected to import/register it themselves.

alenaksu replied:
fixed [6413915](https://github.com/shoelace-style/shoelace/pull/823/commits/641391589df9c9fbd8c598caf6e8e5afe21a1a71)

alenaksu commented:
> Regarding the `lazy` attribute and the `sl-lazy-load` event, I wanted to ask what the difference is between responding to the `sl-lazy-load` event and the plain 'ole `click` event for an item marked as `lazy`? From the [demo](https://shoelace-git-fork-alenaksu-feat-tree-view-shoelace.vercel.app/components/tree?id=lazy-loading) it seems like the lazy loaded children arent loaded until clicking on the `sl-tree-item[lazy]`, right?

Using a `click` event (or maybe `expand`), you would need to prevent the expation of the node, set the `loading` state to true, load the data, set `loading` back to false, set `expand` to true.
So it would look like something like this:
```js
const treeItem = document.querySelector('sl-tree-item.lazy');
let loaded = true;
treeItem.addEventListener('sl-expand', e => {
    if (!loaded) return;
    e.preventDefault();

    treeItem.loading = true;
    // Load data and append the child nodes
    loaded = true;
    treeItem.loading = false;
    treeItem.expand = true;
});
```
With the current implementation instead, you don't nede to take care of this orchestration, you just have to listen for `sl-lazy-load` event and append the child nodes when ready, the rest will be handled automatically.

```js
const treeItem = document.querySelector('sl-tree-item[lazy]');
treeItem.addEventLister('sl-lazy-load', () => {
    // Load data and append the child nodes

    // disable lazy load once done
    treeItem.lazy = false;
});
```
>Also another clarifying question, since I dont see it in the preview docs. Does lazy only apply with selection="none" type trees? (ie, no checkboxes only carets) There's not an example of lazy loading children into a tree with checkboxes, so I wanted to ask if its possible to lazy load children of a selectable item and what that does for the selection aspect? If you can lazy load children of an item with a checkbox, does the checkbox check itself after children are loaded? Do all the children check themselves also?

This part is still in development, but I think that the lazy mode should work with all the types of selection. For what regards the `multiple` selection, I think that if you select a parent node that will lazy load its content, once loaded, it should be select as well, so the selection is delayed until all content gets loaded. 
@claviska what do you think about this scenario?

> I would definitely suggest some more documentation on how and when lazy applies
that's definitely a good suggestion, I'll do it for sure

claviska commented:
> This part is still in development, but I think that the lazy mode should work with all the types of selection.

💯

> For what regards the multiple selection, I think that if you select a parent node that will lazy load its content, once loaded, it should be select as well, so the selection is delayed until all content gets loaded.

This makes sense, that way users can listen for `sl-expand` and know that, whenever it fires, the content is available whether the node is loaded lazily or not.

ralfbecker commented:
I found the attribute-name `multi-select` a bit odd, I believe other attributes are all camelCase.
Maybe even `multiple` makes more sense, as it's used by (sl-)select.

alenaksu commented:
> I found the attribute-name `multi-select` a bit odd, I believe other attributes are all camelCase.
> Maybe even `multiple` makes more sense, as it's used by (sl-)select.

That attribute is going to be removed in favour of the  `selection` attribute (see previous comments)

alenaksu commented:
@break-stuff 
Thank you for taking time to test it, all the accessibility issues should be solved. Below a couple of notes:

> I would also expect the "space" key to toggle the tree item since "space" and "enter" both work to activate buttons.

I didn't change the behavior to stay consistent with what is decribed in the [ARIA APG](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/#keyboard-interaction-24)

> Each tree item should have aria-level, aria-setsize, and aria-posinset specified so screen readers can announce where users are within the hierarchy.

I haven't added this attributes since are not mandatory bu suggested when the browser is not able to compute these informations, for example in the case the treeitems are not yet present in the DOM.

alenaksu commented:
@claviska 
Fixed your comments and implemented the new `selection` attribute

- `none` (_default_) - The user can't select anything
- `single` - The user can select a single item
- `leaf` - The user can select a single item, only leaf nodes. Clicking on a parent node will expand/collapse the node.
- `multiple`  - The user can select multiple items.

break-stuff commented:
> I didn't change the behavior to stay consistent with what is described in the [ARIA APG](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/#keyboard-interaction-24)

I believe they should have the same behavior. If the row is selectable, both "space" and "enter" should select the row. If it is expandable, they both should toggle the row. If it is both expandable and selectable, they both should select the row and the arrow keys should be used to expand and collapse. Check out their example [here](https://www.w3.org/WAI/ARIA/apg/example-index/treeview/treeview-1/treeview-1a.html).

> I haven't added these attributes since are not mandatory but only suggested when the browser is not able to compute these informations, for example in the case the treeitems are not yet present in the DOM.

It looks like the examples in the docs read fine for Narrator and NVDA on Windows, but VoiceOver does not correctly read the position. I tried adding the aria attributes and it didn't seem to help. 🤷‍♂️

alenaksu committed:
chore: update docs

dennerkrans committed:
added swedish translation (#838)

claviska committed:
update changelog

claviska committed:
Merge branch 'feat/tree-view' of github.com:alenaksu/shoelace into alenaksu-feat/tree-view

claviska committed:
various updates

claviska commented:
Spent some time on this today trying to get it over the finish line. I have a branch with my updates here. It's probably easiest to merge my changes from that branch into this PR.

https://github.com/shoelace-style/shoelace/tree/alenaksu-feat/tree-view

The good news is that it's just about there! Here's what I updated:

- Removed `selection="none"` since there's no obvious use case for a tree that isn't selectable. (Sorry, that was originally a suggestion from me but `single|multiple|leaf` seem sufficient!)
- Added a more obvious hover/selection style. You were right on this one too. 😅
- Added missing icon dependency to `<sl-tree>`
- Fixed RTL styles
- Fixed keyboard support when using RTL
- Various style tweaks
- Renamed the `sl-selected-change` event to `--sl-selection-change`
- Renamed the `--indentation-size` custom property to `--indent-size`.
- Added `--indent-guide-color|offset|style|width` custom properties so indent guides can be more easily set.
- Updated examples in the docs to be more generic
- Added remaining React examples
- Updated the changelog

Here's what's left on the wishlist:

- [ ] Disabled items should accept focus but shouldn't expand or accept selection. (This is the recommended behavior for disabled items, and I recently updated Tabs and Menus to do this.)
- [ ] An example with a plus/minus icon instead of chevrons. But rather than bake in specific icons, it would be nice if the user can configure them.
  - In Breadcrumb, there's a hidden slot called called `separator` that the user can insert their own icon into. The component then clones the node, removes ids, and injects it into each breadcrumb item on slotchange.
  - I don't want this to make things more difficult for you, so I'm more than happy to take this on myself in the next iteration if you're not interesting in tackling it. This is really more of a "nice to have."

And there's one known bug I found:

- [ ] If a child node is expanded, tabbing/keyboard stops working when the parent node collapses. I suspect the roving tab index needs to be updated when a child node is hidden, but I didn't have time to look into it today.
  1. To repro, go to the [first example in this preview](https://shoelace-bzzb25yo3-shoelace.vercel.app/components/tree) and tab into the tree.
  2. Press <kbd>Right</kbd> to expand Deciduous."
  3. Press <kbd>Down</kbd> until you land on "Maple."
  4. Press <kbd>Right</kbd> to expand "Maple."
  5. Leaving "Maple" open, press <kbd>Up</kbd> a couple times to go back to "Deciduous."
  6. Press <kbd>Left</kbd> to collapse "Deciduous."
  7. Now try pressing <kbd>Up</kbd> or <kbd>Down</kbd>. The selection remains unchanged and other items can no longer be selected.

alenaksu committed:
fix: collapsed nodes are still focusable

alenaksu committed:
feat: add customizable icons

* fix a bug focusing collapsed nodes

alenaksu committed:
chore: minor refactor

alenaksu committed:
chore: update docs

alenaksu commented:
Thank you for you changes, I've merged your branch into mine.
I also pushed some other stuff:

- The tree now supports customizable icons for the expand/collapse button
- Disabled items are now focusable but not interactive
- Fixed the bug regarding roving tabindex and hidden nodes

alenaksu committed:
chore: update styles
